src/many-modules/src/_5_data/types.rs

```diff
@@ -1,83 +1,566 @@
+                    CborAny::Tagged(Tag::PosBignum, box CborAny::Bytes(v)) => {
+                        Ok(Self::GaugeInt(BigInt::from_bytes_be(Sign::Plus, v)))
+                    }
+                    CborAny::Tagged(Tag::NegBignum, box CborAny::Bytes(v)) => {
+                        Ok(Self::GaugeInt(BigInt::from_bytes_be(Sign::Minus, v)))
+                    }
+                    _ => Err(ManyError::deserialization_error(
+                        "Expected unsigned integer value.",
+                    )),
+                },
+            },
+            DataValue::Counter(v) => Ok(Self::Counter(*v)),
+            DataValue::GaugeInt(bi) => Ok(Self::GaugeInt(bi.clone())),
+        }
+    }
+}
+
+impl<C> Encode<C> for DataValue {
+    fn encode<W: encode::Write>(
+        &self,
+        e: &mut Encoder<W>,
+        _ctx: &mut C,
+    ) -> Result<(), encode::Error<W::Error>> {
+        match self {
+            DataValue::Counter(v) => {
+                e.u64(*v)?;
+            }
+            DataValue::GaugeInt(v) => {
+                if let Some(v) = v.to_i64() {
+                    e.i64(v)?;
+                } else {
+                    match v.to_bytes_be() {
+                        (Sign::Plus, bytes) => {
+                            e.tag(Tag::PosBignum)?.bytes(&bytes)?;
+                        }
+                        (Sign::Minus, bytes) => {
+                            e.tag(Tag::NegBignum)?.bytes(&bytes)?;
+                        }
+                        (Sign::NoSign, _) => {
+                            e.tag(Tag::PosBignum)?.bytes(&[])?;
+                        }
+                    };
+                }
+            }
+            DataValue::Unknown(v) => {
+                e.encode(v)?;
             }
         }
+
+        Ok(())
     }
 }
 
-fn decode_bigint<C>(
-    d: &mut minicbor::Decoder<'_>,
-    _: &mut C,
-) -> Result<BigInt, minicbor::decode::Error> {
-    let vec: Vec<u8> = d.decode()?;
-    Ok(BigInt::from_signed_bytes_be(vec.as_slice()))
+impl<'b> Decode<'b, ()> for DataValue {
+    fn decode(d: &mut Decoder<'b>, _: &mut ()) -> Result<Self, decode::Error> {
+        Ok(DataValue::Unknown(d.decode()?))
+    }
 }
 
-fn encode_bigint<C, W: minicbor::encode::Write>(
-    v: &BigInt,
-    e: &mut minicbor::Encoder<W>,
-    _: &mut C,
-) -> Result<(), minicbor::encode::Error<W::Error>> {
-    e.encode(v.to_signed_bytes_be())?;
-    Ok(())
+impl<'b> Decode<'b, DataType> for DataValue {
+    fn decode(d: &mut Decoder<'b>, ty: &mut DataType) -> Result<Self, decode::Error> {
+        Ok(match ty {
+            DataType::Counter => Self::Counter(d.u64()?),
+            DataType::GaugeInt => Self::GaugeInt(match d.datatype()? {
+                Type::U8
+                | Type::U16
+                | Type::U32
+                | Type::U64
+                | Type::I8
+                | Type::I16
+                | Type::I32
+                | Type::I64 => d.i64()?.into(),
```

Comment:
It's not safe to read an `u64` as an `i64`

u64::MAX : 18_446_744_073_709_551_615u64
i64::MAX : 9_223_372_036_854_775_807i64